trinity/main.py

```diff
@@ -175,5 +170,22 @@ class DBManager(BaseManager):
     for sig in [signal.SIGINT, signal.SIGTERM]:
         loop.add_signal_handler(sig, chain.cancel_token.trigger)
 
-    loop.run_until_complete(run_lightchain(chain))
+    # setup the JSON-RPC server
+    rpc = RPCServer(chain)
+    asyncio.ensure_future(asyncio.start_unix_server(
+        connection_handler(rpc.execute),
+        chain_config.json_rpc_ipc_path,
+        loop=loop,
+        limit=MAXIMUM_REQUEST_BYTES,
+    ), loop=loop)
```

Comment:
I guess this is where the `CancelToken` would be needed? I believe we could move this into a coroutine method under `RPCServer`, where we could do something like

```
async def run(self):
    self.srv = await asyncio.start_unix_server(...)
    await self.cancel_token.wait()
    self.srv.close()

async def wait_closed(self):
    await self.srv.wait_closed()
```

Then `run_chain()` would become something like:

```
async def run_chain():
    try:
        asyncio.ensure_future(chain.peer_pool.run())
        asyncio.ensure_future(rpc.run())
        await chain.run()
    finally:
        ...
        rpc.cancel_token.trigger()
        await rpc.wait_closed()
```
Or, if you don't want to _infect_ the `RPCServer` with asyncio code, I guess those could be standalone functions here

Reply:
Oh, I just realized you'll probably also need to pass the `CancelToken` to `connection_handler()` and from there to `connection_loop()`, where you'd then use `wait_with_token()` and catch `OperationCancelled` to cleanly terminate pending coroutines when we stop the RPC server

Reply:
Good catch, will do.